ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+              commit(false);
+              compactionLog.markCleanupStart();
+              waitTillRefCountZero();
+              cleanup(false);
+              compactionLog.markCycleComplete();
+            }
+            break;
+          default:
+            throw new IllegalStateException("Illegal compaction phase: " + phase);
+        }
+      }
+      endCompaction();
+    } catch (InterruptedException | IOException e) {
+      throw new StoreException("Exception during compaction", e, StoreErrorCodes.Unknown_Error);
+    } finally {
+      runningLatch.countDown();
+    }
+  }
+
+  /**
+   * Checks the sanity of the provided {@code details}
+   * @param details the {@link CompactionDetails} to use for compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   */
+  private void checkSanity(CompactionDetails details) {
+    List<String> segmentsUnderCompaction = details.getLogSegmentsUnderCompaction();
+    LogSegment lastSegment = null;
+    // all segments should be available
+    for (String segmentName : segmentsUnderCompaction) {
+      lastSegment = srcLog.getSegment(segmentName);
+      if (lastSegment == null) {
+        throw new IllegalArgumentException(segmentName + " does not exist in the log");
+      }
+    }
+    // last offset to compact should be outside the range of the journal
+    Offset lastSegmentEndOffset = new Offset(lastSegment.getName(), lastSegment.getEndOffset());
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
+      throw new IllegalArgumentException("Some of the offsets provided for compaction are within the jounral");
+    }
+  }
+
+  /**
+   * Copies valid data from the source log segments into the swap spaces.
+   * <p/>
+   * Splits the compaction into two cycles if there aren't enough swap spaces and completes the copy for the current
+   * cycle.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void copy() throws InterruptedException, IOException, StoreException {
+    setupState();
+    List<String> logSegmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+    Offset startOffsetOfLastIndexSegmentForDelete = getStartOffsetOfLastIndexSegmentForDelete();
+    FileSpan duplicateSearchSpan = null;
+    if (compactionLog.getCurrentIdx() > 0) {
+      // only the very first log segment in the cycle can contain duplicate records.
+      LogSegment logSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(logSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the previous log segment.
```

Comment:
This comment is contradicting the one on line 297. Could you reword? (I know the difference).